src/platform_impl/linux/x11/monitor.rs

```diff
@@ -284,22 +284,19 @@ impl XConnection {
 
     pub fn available_monitors(&self) -> Result<Vec<MonitorHandle>, X11Error> {
         let mut monitors_lock = self.monitor_handles.lock().unwrap();
+        // Copy the cache contents if they are non-empty (Some) producing Result<_, ()>
         (*monitors_lock)
-            .as_ref()
-            .cloned()
-            .map(Ok)
-            .or_else(|| {
-                self.query_monitor_list()
-                    .map(|mon_list| {
-                        let monitors = Some(mon_list);
-                        if !DISABLE_MONITOR_LIST_CACHING {
-                            (*monitors_lock) = monitors.clone();
-                        }
-                        monitors
-                    })
-                    .transpose()
+            .clone()
+            .ok_or(())
```

Comment:
Not sure if `ok_or(())` is the cleanest way of handling this combinator

Reply:
I'm open for suggestions. I looked at https://doc.rust-lang.org/std/option/enum.Option.html searching for `-> Result` and the options are `ok_or`, `ok_or_else`, `transpose` (not really applicable here), `Debug::fmt` (not applicable), `TryFrom` and `TryInto` blanket implementations (which do not do anything useful here).

So, the alternative that I see is a hand-written `match` (which is actually a big win for readability, I think).
```diff
diff --git a/src/platform_impl/linux/x11/monitor.rs b/src/platform_impl/linux/x11/monitor.rs
index e56f54b5..a44e5458 100644
--- a/src/platform_impl/linux/x11/monitor.rs
+++ b/src/platform_impl/linux/x11/monitor.rs
@@ -284,19 +284,16 @@ impl XConnection {
 
     pub fn available_monitors(&self) -> Result<Vec<MonitorHandle>, X11Error> {
         let mut monitors_lock = self.monitor_handles.lock().unwrap();
-        // Copy the cache contents if they are non-empty (Some) producing Result<_, ()>
-        (*monitors_lock)
-            .clone()
-            .ok_or(())
-            // If the cache was empty, actually send a query
-            .or_else(|()| {
-                self.query_monitor_list().map(|mon_list| {
-                    if !DISABLE_MONITOR_LIST_CACHING {
-                        (*monitors_lock) = Some(mon_list.clone());
-                    }
-                    mon_list
-                })
-            })
+        match *monitors_lock {
+            Some(ref monitors) => Ok(monitors.clone()),
+            None => {
+                let monitors = self.query_monitor_list()?;
+                if !DISABLE_MONITOR_LIST_CACHING {
+                    *monitors_lock = Some(monitors.clone());
+                }
+                Ok(monitors)
+            }
+        }
     }
 
     #[inline]
```

Whatya think? (This patch makes the code even three lines shorter...)